R/slice-chop.R

```diff
@@ -0,0 +1,89 @@
+#' Chopping
+#'
+#' @description
+#' - `vec_chop()` provides an efficient method to repeatedly slice a vector. It
+#' captures the pattern of `map(indices, vec_slice, x = x)`.
+#'
+#' - `vec_unchop()` combines a list of vectors into a single vector, placing
+#' elements in the output according to the locations specified by `indices`. It
+#' is similar to [vec_c()], but gives greater control over how the elements
+#' are combined, and does not respect outer names on the list.
```

Comment:
Maybe also connect explicltly with `split()` and `unsplit()`? Should we also provide `vec_unsplit()`?

Reply:
I'm not sure `vec_unsplit()` would be worth it. I think it would have to make too many assumptions about the input? And it is less efficient because it would have to compute `vec_group_loc()` twice.

``` r
library(vctrs)

vec_unsplit <- function(x, by) {
  # - needs to ensure `x` has key and val cols
  # - assume val col is a list of identically typed objects
  # - assumes `by` is the same value that was used to split by
  # - computes `vec_group_loc()` twice (one other time in vec_split())
  
  indices <- vec_group_loc(by)$loc
  vec_unchop(x$val, indices)
}

x <- c(1.5, 2.5, 1, 3, 5)
by <- c("a", "b", "a", "a", "b")

x_split <- vec_split(x, by)

vec_unsplit(x_split, by)
#> [1] 1.5 2.5 1.0 3.0 5.0
```

I think that the chop-apply-unchop approach that we get here might be a better generalization. Using the `ave2()` that is in the examples:

``` r
library(vctrs)

ave2 <- function(.x, .by, .f, ...) {
  # vec_split()
  indices <- vec_group_loc(.by)$loc
  chopped <- vec_chop(.x, indices)
  
  # do stuff
  out <- lapply(chopped, .f, ...)
  
  # vec_unsplit()
  vec_unchop(out, indices)
}

x <- c(1.5, 2.5, 1, 3, 5)
by <- c("a", "b", "a", "a", "b")

ave2(x, by, identity)
#> [1] 1.5 2.5 1.0 3.0 5.0
```